stock_restrict_lot/tests/test_restrict_lot.py

```diff
@@ -0,0 +1,227 @@
+                "product_uom": self.product.uom_id.id,
+                "name": "test",
+                "procure_method": "make_to_stock",
+                "warehouse_id": self.warehouse.id,
+                "route_ids": [(6, 0, self.warehouse.delivery_route_id.ids)],
+                "restrict_lot_id": self.lot.id,
+            }
+        )
+        move._action_confirm()
+        vals_list = move._split(1)
+        new_move = self.env["stock.move"].create(vals_list)
+        self.assertEqual(new_move.restrict_lot_id.id, move.restrict_lot_id.id)
+        other_move = move.copy()
+        self.assertFalse(other_move.restrict_lot_id.id)
+
+    def _update_product_stock(self, qty, lot_id=False, location=None):
+        quant = self.env["stock.quant"].create(
+            {
+                "product_id": self.product.id,
+                "location_id": (
+                    location.id if location else self.warehouse.lot_stock_id.id
+                ),
+                "lot_id": lot_id,
+                "inventory_quantity": qty,
+            }
+        )
+        quant.action_apply_inventory()
+
+    def test_02_move_restrict_lot_reservation(self):
+        lot2 = self.env["stock.lot"].create(
+            {
+                "name": "lot2",
+                "product_id": self.product.id,
+                "company_id": self.warehouse.company_id.id,
+            }
+        )
+        self._update_product_stock(1, lot2.id)
+
+        move = self.env["stock.move"].create(
+            {
+                "product_id": self.product.id,
+                "location_id": self.warehouse.lot_stock_id.id,
+                "location_dest_id": self.customer_loc.id,
+                "product_uom_qty": 1,
+                "product_uom": self.product.uom_id.id,
+                "name": "test",
+                "warehouse_id": self.warehouse.id,
+                "restrict_lot_id": self.lot.id,
+            }
+        )
+        move._action_confirm()
+        move._action_assign()
+        # move should not reserve wrong free lot
+        self.assertEqual(move.state, "confirmed")
+
+        self._update_product_stock(1, self.lot.id)
+        move._action_assign()
+        self.assertEqual(move.state, "assigned")
+        self.assertEqual(move.move_line_ids.lot_id.id, self.lot.id)
+
+    def test_procurement_with_2_steps_output(self):
+        # make warehouse output in two step
+        self.env["res.config.settings"].write(
+            {
+                "group_stock_adv_location": True,
+                "group_stock_multi_locations": True,
+            }
+        )
+        warehouse = self.env["stock.warehouse"].search(
+            [("company_id", "=", self.env.company.id)], limit=1
+        )
+        warehouse.delivery_steps = "pick_ship"
+
+        self.product.categ_id.route_ids |= self.env["stock.route"].search(
+            [("name", "ilike", "deliver in 2")]
+        )
+        # self.env["stock.warehouse"].write(dict(delivery_steps='pick_ship',))
```

Comment:
just realise I've let this useless line while coding this test on V14.0...

Reply:
Let's remove it then !